merkle-tree/concurrent/src/lib.rs

```diff
@@ -0,0 +1,360 @@
+    ///     the first element of the proof).
+    fn update_proof(
+        &self,
+        changelog_index: usize,
+        leaf_index: usize,
+        proof: &[[u8; 32]; MAX_HEIGHT],
+    ) -> Result<[[u8; 32]; MAX_HEIGHT], HasherError> {
+        let mut updated_proof = proof.to_owned();
+        let mut k = changelog_index;
+
+        while k != self.current_changelog_index as usize {
+            let changelog_entry = self.changelog[k];
+            let changelog_entry_index = changelog_entry.index as usize;
+            if leaf_index != changelog_entry_index {
+                let intersection_index = self.intersection_index(leaf_index, changelog_entry_index);
+                updated_proof[intersection_index] = changelog_entry.path[intersection_index];
+            } else {
+                return Err(HasherError::CannotUpdateLeaf);
+            }
+
+            k = (k + 1) % MAX_ROOTS;
+        }
+
+        Ok(updated_proof)
+    }
+
+    /// Updates the leaf under `leaf_index` with the `new_leaf` value.
+    ///
+    /// # Validation
+    ///
+    /// This method doesn't validate the proof. Caller is responsible for
+    /// doing that before.
+    fn update_leaf_in_tree(
+        &mut self,
+        new_leaf: &[u8; 32],
+        leaf_index: usize,
+        proof: &[[u8; 32]; MAX_HEIGHT],
+    ) -> Result<(), HasherError> {
+        let mut node = *new_leaf;
+        let mut changelog_path = [[0u8; 32]; MAX_HEIGHT];
+
+        for (j, sibling) in proof.iter().enumerate() {
+            changelog_path[j] = node;
+            node = compute_parent_node::<H>(&node, sibling, leaf_index, j)?;
+        }
+
+        let changelog_entry = ChangelogEntry::new(node, changelog_path, leaf_index);
+        self.inc_current_changelog_index();
+        if let Some(changelog_element) = self
+            .changelog
+            .get_mut(self.current_changelog_index as usize)
+        {
+            *changelog_element = changelog_entry
+        }
+
+        self.inc_current_root_index();
+        *self
+            .roots
+            .get_mut(self.current_root_index as usize)
+            .ok_or(HasherError::RootsZero)? = node;
+
+        if self.rightmost_index > 0 && leaf_index == self.rightmost_index as usize - 1 {
+            self.rightmost_proof.copy_from_slice(proof);
+            self.rightmost_leaf = *new_leaf;
+        }
+
+        Ok(())
+    }
+
+    /// Replaces the `old_leaf` under the `leaf_index` with a `new_leaf`, using
+    /// the given `proof` and `changelog_index` (pointing to the changelog entry
+    /// which was the newest at the time of preparing the proof).
+    pub fn update(
+        &mut self,
+        changelog_index: usize,
+        old_leaf: &[u8; 32],
+        new_leaf: &[u8; 32],
+        leaf_index: usize,
+        proof: &[[u8; 32]],
+    ) -> Result<(), HasherError> {
+        // Provided proof might have lower number of elements than the height.
```

Comment:
Why do we do that?
Is there an example where we can be sure that we will always send a lower number of elements?
Because if we are not certain that we will always send a lower number of elements we will need to consider the worst case in the Solana program thus cannot use the benefit that we can send less data in some edge cases.

My first thought was that this might be an attack vector, thus if we don't need it why not eliminate this option?

Reply:
I think it would be cleaner to just require a constant proof size, thus replace ``proof: &[[u8; 32]],`` with ``        proof: &[[u8; 32]; MAX_HEIGHT],``

Reply:
OK, after giving it some thought, I agree - I think enforcing the full proof size would be more secure.

Just to explain myself what was done here:

In the current implementation and with how tests are done, the "not full" proof will be always lower when the tree is not filled up. It's just for the convenience, so the user doesn't have to include zero bytes in the proof.

Let's imagine this tree:

```
              R
           /     \
          H4    Z[3]
        /    \
      H3    Z[2]
    /    \
  H1      H2
 /  \    /  \
L1  L2  L3  L4
```

When you update L1, the full proof is:

```
[L2, H2, H::zero_bytes()[2], H::zero_bytes()[3]]
```

An the "not full" proof, without zero bytes is:

```
[L2, H2]
```

https://github.com/Lightprotocol/light-protocol/blob/e8f400c5f4fd4c4e1a888f084315b8ba1981541c/merkle-tree/concurrent/tests/tests.rs#L276-L282

Reply:
But anyway, I applied your idea :)

Reply:
Thanks for clarifying and applying it!

In case we decide that this is worth optimizing in the future we can still write a wrapper which tries to add missing elements later. Thinking about it there might be potential here but it's not critical now and I think it makes sense to have a very strict check at this level of the implementation.